### PR TITLE
wallet: Upgrade path for non-HD wallets to HD

### PR DESCRIPTION
**Pull request by KolbyML**
This PR sets to solve issue #3740 

It creates an upgrade path for people with non-HD wallets to upgrade to an HD wallet

It encompasses two parts
1. the split keypool code from bitcoin/bitcoin#12560
2. code to allow you to upgrade to HD which I did

Unlike bitcoins solution of upgrading on boot. I went with upgrading through an RPC command. This was because the only way to upgrade an encrypted wallet was with a password. Unless you passed your password into the daemon you would only be able to upgrade non-encrypted wallets, which bitcoin has not supported, which makes their implementation kind of pointless, as most people would encrypt their wallet.

Of course, there could be other upgrade methods possible such as through the GUI in the future, if people wanted that.

### CODE REVIEW

File: src/keystore.cpp [hunk 1]
```diff
@@ -149,3 +149,10 @@ bool CBasicKeyStore::GetHDChain(CHDChain& hdChainRet) const
     hdChainRet = hdChain;
     return !hdChain.IsNull();
 }
+
+bool HaveKey(const CKeyStore& store, const CKey& key)
+{
+    CKey key2;
+    key2.Set(key.begin(), key.end(), !key.IsCompressed());
+    return store.HaveKey(key.GetPubKey().GetID()) || store.HaveKey(key2.GetPubKey().GetID());
+}
```
Comment (PastaPastaPasta) [inline src/keystore.cpp:158]:
```suggestion
}

```

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

**Comment by UdjinM6** [comment]
I like the approach (upgrade via rpc and not via cm-line param) 👍  But the code here has too many issues unfortunately:
- there is no difference between HD and HD split versions in Dash (our HD is already the one with the split), all the split logic added doesn't really apply here,
- new `HaveKey` isn't used,
- chain assignment in `EncryptHDChain` should happen after other checks,
- `generatenewseedwords` shouldn't be needed because that's what an empty mnemonic already does,
- some `HelpExampleCli`s are wrong,
- some more redundant code.

I tried to implement patches to fix these issues locally and realized that I basically had to rewrite most of the code 🙈  So I went ahead and simply reworked the whole thing, pls see #3933.

**Comment by KolbyML** [comment]
Closing this PR, cause @UdjinM6 made a far more improved PR, I think I can learn a lot from it and hopefully, I can do solve problems that good one day.

**Comment by UdjinM6** [comment]
Thank you for looking into issues and for contributing! 👍

**Comment by KolbyML** [comment]
> Thank you for looking into issues and for contributing! 👍

Thank you, hopefully, I can do it again one day and make a better PR. I think your PR can teach me a lot.

**Pull request closed without merging**